prov/shm/src/smr_dsa.c

```diff
@@ -132,32 +142,41 @@ struct dsa_ops {
 static void *libdsa_handle = NULL;
 static struct dsa_ops dsa_ops;
 
-static inline unsigned char dsa_enqcmd(struct dsa_hw_desc *desc,
-				       volatile void *reg)
+static int dsa_enq_cmd(struct wq_handle *wq_handle, struct dsa_hw_desc *desc)
 {
+	volatile void *reg = (void *) wq_handle->mmapped; 
 	unsigned char retry;
 
 	asm volatile(".byte 0xf2, 0x0f, 0x38, 0xf8, 0x02\t\n"
 		     "setz %0\t\n"
 		     : "=r"(retry)
 		     : "a"(reg), "d"(desc));
-	return retry;
+	return (int) retry;
+}
+
+static int dsa_write_cmd(struct wq_handle *wq_handle, struct dsa_hw_desc *desc)
+{
+	int ret;
+	int wq_fd = wq_handle->fd;
+
+	ret = write(wq_fd, desc, sizeof(*desc));
+	return ret != sizeof(*desc) ? -1 : 0;
 }
 
 static __always_inline void dsa_desc_submit(struct smr_dsa_context *dsa_context,
 					    struct dsa_hw_desc *hw)
 {
-	int enq_status;
+	int status;
 
 	// make sure writes (e.g., comp.status = 0) are ordered wrt to enqcmd
 	{ asm volatile("sfence":::"memory"); }
 
 	do {
-		enq_status = dsa_enqcmd(hw,
-				dsa_context->wq_portal[dsa_context->next_wq]);
-		dsa_context->next_wq = (dsa_context->next_wq + 1) %
-			(dsa_context->wq_count);
-	} while (enq_status);
+		status = dsa_context->submit_cmd(
+			&dsa_context->wq_handle[dsa_context->next_wq], hw);
+		dsa_context->next_wq =
+			(dsa_context->next_wq + 1) % (dsa_context->wq_count);
+	} while (status);
```

Comment:
enq_cmd returns 1 as status when the status is retry. write_cmd function as written returns -1 as status when the status is retry. Not sure if the loop will work correctly if status is -1. Change the write_cmd function to also return 1 in case of retry.